### PR TITLE
Update Rancher Machine CLI invocations

### PR DESCRIPTION
**Pull request by bfbachmann**
## Problem
The original problem this PR (in combination with https://github.com/rancher/machine/pull/224) addresses is the following
1. Create a new node-provisioned cluster (RKE1 or 2, doesn't matter)
2. After the cluster has been created, rotate the associated cloud credential (i.e. change it in the cloud provider and update it in the Rancher UI)
3. Try to delete the cluster. You'll notice that the nodes cannot be deleted even though Rancher has the most up-to-date cloud credential.

I solved this problem in Rancher machine by simply having Rancher machine look for and use the updated cloud credential if the `--update-config` flag is passed. Rancher Machine will still behave as it always has if that flag is not passed.
 
## Solution
This PR just updates Rancher to pass the `--update-config` flag on the `rancher-machine rm` command so Rancher Machine knows to re-load configuration from envvars (for RKE2) or CLI flags (for RKE1).
 
### Manual Testing
- Create RKE1 and 2 node-provisioned clusters
- Rotate credentials
- Try delete the clusters and ensure that all nodes are properly deleted

### Automated Testing

* Test types added/modified:
    * Unit: Included in https://github.com/rancher/machine/pull/224/files#diff-835f2809e3a2abc47591f343924d8981dc259d95f1f5eddeef8c7543cf5368ffR39
    * Integration: https://github.com/rancher/machine/pull/224/files#diff-835f2809e3a2abc47591f343924d8981dc259d95f1f5eddeef8c7543cf5368ffR39
  
Summary: I wrote unit and full CLI tests in my Rancher Machine PR to ensure that the CLI still works as intended.

### Regressions Considerations

I think the main thing to focus on is the case where credentials are changed after cluster creation. This is relevant for all node drivers, so we probably want to test this scenario with each one to ensure it works as intended.

Existing / newly added automated tests that provide evidence there are no regressions:

### CODE REVIEW

File: pkg/controllers/capr/machineprovision/args.go [hunk 1]
```diff
@@ -143,7 +143,13 @@ func (h *handler) getArgsEnvAndStatus(infra *infraObject, args map[string]interf
...
 		cmd = append(cmd, toArgs(driver, args, rancherCluster.Status.ClusterName)...)
 	} else {
-		cmd = append(cmd, "rm", "-y")
+		// We're passing the `--update-config` flag here along with driver-specific flags to tell Rancher Machine
+		// to reload the driver-specific flags we pass via envvars. They need to be reloaded to account for cases
+		// were configuration used by Rancher machine (namely cloud credentials) change while the machine is still running.
+		// If we didn't do this, a user could create a machine and change the cloud credential it uses, leaving Rancher
+		// unable to remove the machine afterward because Rancher machine will always use the old credential it stored on
+		// machine creation.
+		cmd = append(cmd, "rm", "-y", "--update-config")
```
Comment (jiaqiluo) [inline pkg/controllers/capr/machineprovision/args.go:152]: The comment says we need to append the driver-specific flags here. Are they missed?

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: pkg/controllers/capr/machineprovision/args.go (hunk 1 above)
Comment (bfbachmann) [reply pkg/controllers/capr/machineprovision/args.go:152]: They're included in envvars which are populated with `envSecret` in this function.

File: pkg/controllers/management/node/utils.go [hunk 2]
```diff
@@ -67,8 +67,20 @@ func buildCreateCommand(node *v3.Node, configMap map[string]interface{}) []strin
...
+	cmd = append(cmd, buildDriverFlags(sDriver, configMap)...)
+
+	// Add the hostname to the command.
+	cmd = append(cmd, node.Spec.RequestedHostname)
+	return cmd
+}
+
+// buildDriverFlags extracts driver-specific configuration from the given configmap and turns it into CLI flags.
+func buildDriverFlags(driverName string, configMap map[string]interface{}) []string {
+	cmd := make([]string, 0)
```
Comment (jiaqiluo) [inline pkg/controllers/management/node/utils.go:80]: nit.

```suggestion
	flags := make([]string, len(configMap))
```

**Review by jiaqiluo — CHANGES_REQUESTED** [review]
Overall LGTM.
Please squash commits into one.

File: pkg/controllers/management/node/utils.go (hunk 2 above)
Comment (bfbachmann) [reply pkg/controllers/management/node/utils.go:80]: Based on the code, I don't think `cmd` will necessarily have the same number of entries as the configmap, so this wouldn't work.

Comment (jiaqiluo) [reply pkg/controllers/management/node/utils.go:80]: Ok, it makes sense.

**Review by jiaqiluo — APPROVED** [review]
LGTM 💯

Please squash commits into one & rebase against the upstream before merging.

File: pkg/controllers/management/node/controller.go [hunk 3]
```diff
@@ -275,7 +275,18 @@ func (m *Lifecycle) Remove(machine *apimgmtv3.Node) (obj runtime.Object, err err
...
 				return machine, err
 			}
-			if err = deleteNode(config.Dir(), machine); err != nil {
+
+			driverConfig, err := config.DriverConfig()
+			if err != nil {
+				return nil, err
+			}
+
+			configRawMap := map[string]interface{}{}
```
Comment (jakefhyde) [inline pkg/controllers/management/node/controller.go:284]:
```suggestion
			configRawMap := map[string]any{}
```

File: pkg/controllers/management/node/utils.go [hunk 4]
```diff
@@ -290,8 +301,18 @@ func nodeExists(nodeDir string, node *v3.Node) (bool, error) {
 	return false, nil
 }
 
-func deleteNode(nodeDir string, node *v3.Node) error {
-	command, err := buildCommand(nodeDir, node, []string{"rm", "-f", node.Spec.RequestedHostname})
+func deleteNode(nodeDir string, node *v3.Node, configMap map[string]interface{}) error {
```
Comment (jakefhyde) [inline pkg/controllers/management/node/utils.go:304]:
```suggestion
func deleteNode(nodeDir string, node *v3.Node, configMap map[string]any) error {
```

File: pkg/controllers/management/node/utils.go [hunk 5]
```diff
@@ -67,8 +67,20 @@ func buildCreateCommand(node *v3.Node, configMap map[string]interface{}) []strin
...
+	// Append driver-specific flags to the command.
+	cmd = append(cmd, buildDriverFlags(sDriver, configMap)...)
+
+	// Add the hostname to the command.
+	cmd = append(cmd, node.Spec.RequestedHostname)
+	return cmd
+}
+
+// buildDriverFlags extracts driver-specific configuration from the given configmap and turns it into CLI flags.
+func buildDriverFlags(driverName string, configMap map[string]interface{}) []string {
```
Comment (jakefhyde) [inline pkg/controllers/management/node/utils.go:79]:
```suggestion
func buildDriverFlags(driverName string, configMap map[string]any) []string {
```

**Review by jakefhyde — CHANGES_REQUESTED** [review]

**Review by thatmidwesterncoder — DISMISSED** [review]
Looks good after Jake's suggestions.

File: pkg/api/norman/store/cluster/cluster_store_test.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package cluster
```
Comment (jakefhyde) [inline pkg/api/norman/store/cluster/cluster_store_test.go:1]: This file appears to have been added accidentally

**Review by jakefhyde — CHANGES_REQUESTED** [review]

**Review by jakefhyde — APPROVED** [review]

**Review by jiaqiluo — APPROVED** [review]
LGMT.
Please squash commits before merging.

**Pull request merged**